File: tests/test_maglit.cpp
#include <cstring>
#include <gtest/gtest.h>

#include "../maglit/m3dc1_source.h"
#include "../maglit/maglit.h"

// Test fixture for maglit functionality
class MaglitTest : public ::testing::Test {
  protected:
    static void SetUpTestSuite() {
        std::snprintf(source_path, sizeof(source_path),
                      "%s/C1.h5", TEST_DATA_DIR);
        shape_path  = std::string(TEST_DATA_DIR) + "/tcabr_first_wall.txt";
        source_vac  = new M3DC1Source(source_path, -1);
        tracer_vac  = new maglit(*source_vac);
        tracer_vac->configure(0.01, 1e-6, 0.1);
        source_resp = new M3DC1Source(source_path, 1);
        tracer_resp = new maglit(*source_resp);
        tracer_resp->configure(0.01, 1e-6, 0.1);
    }

    static void TearDownTestSuite() {
        // Delete tracers before sources (tracers hold references to sources).
        delete tracer_vac;
        delete tracer_resp;
        delete source_vac;
        delete source_resp;
        tracer_vac  = nullptr;
        tracer_resp = nullptr;
        source_vac  = nullptr;
        source_resp = nullptr;
    }

    void SetUp() override { tolerance = 1e-6; }

    double              tolerance;
    static char         source_path[256];
    static std::string  shape_path;
    static M3DC1Source *source_vac;
    static M3DC1Source *source_resp;
    static maglit      *tracer_vac;
    static maglit      *tracer_resp;

    bool values_are_close(double a, double b, double tol) {
        return std::fabs(a - b) < tol;
    }
};

// Static member definitions
M3DC1Source *MaglitTest::source_vac  = nullptr;
M3DC1Source *MaglitTest::source_resp = nullptr;
maglit      *MaglitTest::tracer_vac  = nullptr;
maglit      *MaglitTest::tracer_resp = nullptr;
char         MaglitTest::source_path[256];
std::string  MaglitTest::shape_path;

// ==================== MAGLIT INTEGRATION TESTS ====================

// Test constructor and configuration
TEST_F(MaglitTest, ConstructorAndConfiguration) {
    ASSERT_NE(tracer_vac, nullptr);
    EXPECT_TRUE(true); // If we reach here, constructors and configurations worked
}

// Test magnetic field evaluation at a specific point
TEST_F(MaglitTest, MagneticFieldEvaluation) {
    double x[3] = {0.7, 0.0, 0.0}; // R, phi, Z
    double B[3] = {0.0, 0.0, 0.0}; // B_R, B_phi, B_Z

    bool success = tracer_vac->calc_mag_field(x, B);
    ASSERT_TRUE(success);
    // Basic checks on magnetic field values
    // Compared to fusion-io python routines output
    EXPECT_TRUE(values_are_close(B[0], -0.00401528, tolerance)); // B_R
    EXPECT_TRUE(values_are_close(B[1], -0.95980778, tolerance)); // B_phi
    EXPECT_TRUE(values_are_close(B[2], -0.14235661, tolerance)); // B_Z
}

// Test poloidal flux evaluation at a specific point
TEST_F(MaglitTest, PsiFieldEvaluation) {
    double x[3] = {0.7, 0.0, 0.0}; // R, phi, Z
    double psi = 0.0;

    tracer_vac->psi_eval(x[0], x[1], x[2], &psi);
    // Basic checks on psi values
    // Compared to fusion-io python routines output
    EXPECT_TRUE(values_are_close(psi, -0.00988012, tolerance)); // psi
}

// Test normalized poloidal flux evaluation at a specific point
TEST_F(MaglitTest, PsiNFieldEvaluation) {
    double x[3] = {0.7, 0.0, 0.0}; // R, phi, Z
    double psiN = 0.0;

    tracer_vac->psin_eval(x[0], x[1], x[2], &psiN);
    // Basic checks on psiN values
    // Compared to fusion-io python routines output
    EXPECT_TRUE(values_are_close(psiN, 0.328436, tolerance)); // psiN
}

// Test inverse map functionality
TEST_F(MaglitTest, InverseMap) {
    double R = 0.7;
    double Z = 0.0;
    double phi = 0.0;
    double phi_max = M_PI / 2; // 90 degrees

    tracer_vac->reset();
    tracer_vac->inverse_map(false); // Forward map
    int status_fwd = SODE_CONTINUE_GOOD_STEP;
    while (status_fwd == SODE_CONTINUE_GOOD_STEP || status_fwd == SODE_CONTINUE_BAD_STEP) {
        status_fwd = tracer_vac->step(R, Z, phi, phi_max, 0);
    }
    EXPECT_EQ(status_fwd, SODE_SUCCESS_TIME);
    EXPECT_TRUE(values_are_close(phi, phi_max, tolerance));

    tracer_vac->reset();
    tracer_vac->inverse_map(true); // Inverse map
    phi_max += phi;                // Inverting the map does not reset phi nor makes it decrease when iterating
    int status_inv = SODE_CONTINUE_GOOD_STEP;
    while (status_inv == SODE_CONTINUE_GOOD_STEP || status_inv == SODE_CONTINUE_BAD_STEP) {
        status_inv = tracer_vac->step(R, Z, phi, phi_max, 0);
    }

    EXPECT_EQ(status_inv, SODE_SUCCESS_TIME);
    EXPECT_TRUE(values_are_close(R, 0.7, tolerance));
    EXPECT_TRUE(values_are_close(Z, 0.0, tolerance));
}

// ==================== MAGLIT MONITORING TESTS ====================

TEST_F(MaglitTest, ColliderMonitorIntegration) {
    // Initial conditions — point inside the chaotic region
    double last_R = 0;
    double last_Z = 0;
    double R = 0.5;
    double Z = -0.225;
    double phi = 0.0;
    double phi_max = 10000 * 2 * M_PI; // Large phi to ensure crossing

    tracer_resp->reset();
    tracer_resp->set_monitor(shape_path);
    tracer_resp->inverse_map(false); // Forward map

    // Integrate while checking for wall crossings (handled internally by monitor)
    int status = SODE_CONTINUE_GOOD_STEP;
    while (status == SODE_CONTINUE_GOOD_STEP || status == SODE_CONTINUE_BAD_STEP) {
        last_R = R;
        last_Z = Z;
        status = tracer_resp->step(R, Z, phi, phi_max, -1);
    }

    // If collider works, integration should stop due to monitor trigger
    EXPECT_EQ(status, SODE_SUCCESS_MONITOR);
    EXPECT_TRUE(tracer_resp->get_boundary().inside(last_R, last_Z));
    EXPECT_FALSE(tracer_resp->get_boundary().inside(R, Z));
}

TEST_F(MaglitTest, InverseColliderMonitorIntegration) {
    // Initial conditions — point inside the chaotic region
    double last_R = 0;
    double last_Z = 0;
    double R = 0.5;
    double Z = -0.225;
    double phi = 0.0;
    double phi_max = 10000 * 2 * M_PI; // Large phi to ensure crossing

    tracer_resp->reset();
    tracer_resp->set_monitor(shape_path);
    tracer_resp->inverse_map(true); // Inverse map

    // Integrate while checking for wall crossings (handled internally by monitor)
    int status = SODE_CONTINUE_GOOD_STEP;
    while (status == SODE_CONTINUE_GOOD_STEP || status == SODE_CONTINUE_BAD_STEP) {
        last_R = R;
        last_Z = Z;
        status = tracer_resp->step(R, Z, phi, phi_max, -1);
    }

    // If collider works, integration should stop due to monitor trigger
    EXPECT_EQ(status, SODE_SUCCESS_MONITOR);
    EXPECT_TRUE(tracer_resp->get_boundary().inside(last_R, last_Z));
    EXPECT_FALSE(tracer_resp->get_boundary().inside(R, Z));
    tracer_resp->inverse_map(false); // restore default
}

TEST_F(MaglitTest, OnBoundaryColliderMonitorIntegration) {
    // Initial conditions — point on the boundary
    double last_R = 0;
    double last_Z = 0;
    double R = 0.435;
    double Z = -0.235;
    double phi = 0.0;
    double phi_max = 10000 * 2 * M_PI; // Large phi to ensure crossing

    tracer_resp->reset();
    tracer_resp->set_monitor(shape_path);
    tracer_resp->inverse_map(false); // Forward map

    // Integrate while checking for wall crossings (handled internally by monitor)
    int status = SODE_CONTINUE_GOOD_STEP;
    int step_count = 0;
    while (status == SODE_CONTINUE_GOOD_STEP || status == SODE_CONTINUE_BAD_STEP) {
        last_R = R;
        last_Z = Z;
        status = tracer_resp->step(R, Z, phi, phi_max, -1);
        step_count++;
    }

    // If collider works, integration should stop due to monitor trigger
    EXPECT_GT(step_count, 0);
    EXPECT_EQ(status, SODE_SUCCESS_MONITOR);
    EXPECT_TRUE(tracer_resp->get_boundary().inside(last_R, last_Z));
    EXPECT_FALSE(tracer_resp->get_boundary().inside(R, Z));
}

File: tests/python/test_maglit.py
"""
Python tests for M3DC1Source and Maglit.

Mirrors the C++ GoogleTest suite in tests/test_maglit.cpp. Requires
tests/data/C1.h5 and tests/data/tcabr_first_wall.txt.
"""

import math
import os

import numpy as np
import pytest
import maglib
from maglib import M3DC1Source, Maglit, SodeStatus

DATA_DIR = os.path.join(os.path.dirname(__file__), "..", "data")
C1_H5    = os.path.join(DATA_DIR, "C1.h5")
WALL_TXT = os.path.join(DATA_DIR, "tcabr_first_wall.txt")

TOL = 1e-6


# ── Fixtures ──────────────────────────────────────────────────────────────────

@pytest.fixture(scope="module")
def source_vac():
    src = M3DC1Source(C1_H5, -1)
    assert src.is_valid(), "vacuum M3DC1Source failed to open"
    return src


@pytest.fixture(scope="module")
def source_resp():
    src = M3DC1Source(C1_H5, 1)
    assert src.is_valid(), "response M3DC1Source failed to open"
    return src


@pytest.fixture(scope="module")
def tracer_vac(source_vac):
    t = Maglit(source_vac)
    t.configure(0.01, 1e-6, 0.1)
    return t


@pytest.fixture(scope="module")
def tracer_resp(source_resp):
    t = Maglit(source_resp)
    t.configure(0.01, 1e-6, 0.1)
    return t


# ── M3DC1Source ───────────────────────────────────────────────────────────────

def test_m3dc1_construction_valid(source_vac):
    assert source_vac.is_valid()


def test_m3dc1_construction_invalid():
    src = M3DC1Source("/nonexistent/path.h5", -1)
    assert not src.is_valid()


def test_m3dc1_two_timeslices(source_vac, source_resp):
    assert source_vac.is_valid()
    assert source_resp.is_valid()


# ── Maglit construction ───────────────────────────────────────────────────────

def test_maglit_construction(tracer_vac):
    assert tracer_vac is not None


# ── Magnetic field evaluation ─────────────────────────────────────────────────

def test_calc_mag_field(tracer_vac):
    ok, B = tracer_vac.calc_mag_field(R=0.7, phi=0.0, Z=0.0)
    assert ok
    assert isinstance(B, np.ndarray)
    assert B.shape == (3,)
    assert abs(B[0] - (-0.00401528)) < TOL   # B_R
    assert abs(B[1] - (-0.95980778)) < TOL   # B_phi
    assert abs(B[2] - (-0.14235661)) < TOL   # B_Z


def test_calc_mag_field_nonzero(tracer_vac):
    ok, B = tracer_vac.calc_mag_field(0.62, 0.0, 0.0)
    assert ok
    assert np.all(np.isfinite(B))
    assert np.linalg.norm(B) > 0.0


# ── Flux evaluation ───────────────────────────────────────────────────────────

def test_psi_eval(tracer_vac):
    ok, psi = tracer_vac.psi_eval(R=0.7, phi=0.0, Z=0.0)
    assert ok
    assert isinstance(psi, float)
    assert abs(psi - (-0.00988012)) < TOL


def test_psin_eval(tracer_vac):
    ok, psin = tracer_vac.psin_eval(R=0.7, phi=0.0, Z=0.0)
    assert ok
    assert isinstance(psin, float)
    assert abs(psin - 0.328436) < TOL


# ── Step integration ──────────────────────────────────────────────────────────

def _integrate(tracer, R, Z, phi, phi_max, dir=0):
    """Drive step() in a loop until a terminal status is reached."""
    tracer.reset()
    status = SodeStatus.CONTINUE_GOOD
    while status in (SodeStatus.CONTINUE_GOOD, SodeStatus.CONTINUE_BAD):
        R, Z, phi, status = tracer.step(R, Z, phi, phi_max, dir)
    return R, Z, phi, status


def test_step_returns_tuple(tracer_vac):
    tracer_vac.reset()
    result = tracer_vac.step(0.7, 0.0, 0.0, math.pi / 2)
    assert len(result) == 4
    R, Z, phi, status = result
    assert isinstance(R, float)
    assert isinstance(Z, float)
    assert isinstance(phi, float)


def test_step_reaches_phi_max(tracer_vac):
    phi_max = math.pi / 2
    R, Z, phi, status = _integrate(tracer_vac, 0.7, 0.0, 0.0, phi_max)
    assert status == SodeStatus.SUCCESS_TIME
    assert abs(phi - phi_max) < TOL


def test_inverse_map_roundtrip(tracer_vac):
    """Forward half-turn then backward half-turn must return to start."""
    R0, Z0, phi0 = 0.7, 0.0, 0.0
    phi_half = math.pi / 2

    try:
        tracer_vac.inverse_map(False)
        R1, Z1, phi1, status = _integrate(tracer_vac, R0, Z0, phi0, phi_half)
        assert status == SodeStatus.SUCCESS_TIME

        tracer_vac.inverse_map(True)
        phi_back = phi1 + phi_half
        R2, Z2, phi2, status = _integrate(tracer_vac, R1, Z1, phi1, phi_back)
        assert status == SodeStatus.SUCCESS_TIME
        assert abs(R2 - R0) < TOL
        assert abs(Z2 - Z0) < TOL
    finally:
        tracer_vac.inverse_map(False)  # restore default even on failure


# ── Boundary monitor ──────────────────────────────────────────────────────────

def test_set_monitor(tracer_resp):
    tracer_resp.set_monitor(WALL_TXT)


def test_monitor_stops_at_wall(tracer_resp):
    """A field line starting inside the vessel should eventually hit the wall."""
    tracer_resp.set_monitor(WALL_TXT)
    phi_max = 10000 * 2 * math.pi

    last_R = last_Z = 0.0
    R, Z, phi = 0.5, -0.225, 0.0
    tracer_resp.reset()
    tracer_resp.inverse_map(False)

    status = SodeStatus.CONTINUE_GOOD
    while status in (SodeStatus.CONTINUE_GOOD, SodeStatus.CONTINUE_BAD):
        last_R, last_Z = R, Z
        R, Z, phi, status = tracer_resp.step(R, Z, phi, phi_max, -1)

    assert status == SodeStatus.SUCCESS_MONITOR
    assert tracer_resp.boundary.inside(last_R, last_Z)
    assert not tracer_resp.boundary.inside(R, Z)
